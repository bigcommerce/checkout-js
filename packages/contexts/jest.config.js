module.exports = {
    displayName: 'contexts',
    preset: '../../jest.preset.js',
    transform: {
        '^.+\\.[tj]sx?$': ['ts-jest', {
            tsconfig: '<rootDir>/tsconfig.spec.json',
            diagnostics: false,
        }],
    },
    collectCoverageFrom: ['src/**/*.{ts,tsx}', '!src/**/*/index.ts', '!src/icon/*.tsx'],
    setupFilesAfterEnv: ['../../jest-setup.ts'],
    coverageDirectory: '../../coverage/packages/contexts',
};
