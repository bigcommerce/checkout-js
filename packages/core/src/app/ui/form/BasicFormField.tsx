import { Field, FieldConfig, FieldProps, getIn } from 'formik';
import { isDate, noop } from 'lodash';
import React, {
    Component,
    createElement,
    FunctionComponent,
    memo,
    useCallback,
    useMemo,
} from 'react';
import shallowEqual from 'shallowequal';

import FormFieldContainer from './FormFieldContainer';

export interface BasicFormFieldProps extends FieldConfig {
    additionalClassName?: string;
    className?: string;
    testId?: string;
    onChange?(value: any): void;
}

const BasicFormField: FunctionComponent<BasicFormFieldProps> = ({
    additionalClassName,
    className,
    component,
    render,
    testId,
    onChange,
    ...rest
}) => {
    const renderInnerField = useCallback(
        (props: FieldProps) => (
            <InnerField
                {...props}
                additionalClassName={additionalClassName}
                className={className}
                component={component}
                onChange={onChange}
                render={render}
                testId={testId}
            />
        ),
        [additionalClassName, className, component, render, testId, onChange],
    );

    return <Field {...rest} render={renderInnerField} />;
};

type InnerFieldProps = Omit<BasicFormFieldProps, keyof FieldConfig> & InnerFieldInputProps;

const InnerField: FunctionComponent<InnerFieldProps> = memo(
    ({ additionalClassName, component, field, form, onChange, render, testId }) => {
        const input = useMemo(
            () => (
                <InnerFieldInput
                    component={component}
                    field={field}
                    form={form}
                    onChange={onChange}
                    render={render}
                />
            ),
            [field, form, onChange, component, render],
        );

        return (
            <FormFieldContainer
                additionalClassName={additionalClassName}
                hasError={getIn(form.errors, field.name)}
                testId={testId}
            >
                {input}
            </FormFieldContainer>
        );
    },
    (
        { form: prevForm, field: prevField, ...prevProps },
        { form: nextForm, field: nextField, ...nextProps },
    ) =>
        shallowEqual(prevProps, nextProps) &&
        shallowEqual(prevForm, nextForm) &&
        shallowEqual(prevField, nextField),
);

type InnerFieldInputProps = FieldProps &
    Pick<FieldConfig, 'component' | 'render'> & {
        onChange?(value: string): void;
    };

class InnerFieldInput extends Component<InnerFieldInputProps> {
    componentDidUpdate({ field: prevField }: InnerFieldInputProps) {
        const {
            field: { value },
            onChange = noop,
        } = this.props;
        const comparableValue = isDate(value) ? value.getTime() : value;
        const comparablePrevValue = isDate(prevField.value)
            ? prevField.value.getTime()
            : prevField.value;

        if (comparableValue !== comparablePrevValue) {
            onChange(value);
        }
    }

    render() {
        const { component = 'input', field, render } = this.props;

        if (render) {
            return (render as any)(this.props);
        }

        if (typeof component === 'string') {
            return createElement(component as any, field);
        }

        return createElement(component as any, this.props);
    }
}

export default memo(BasicFormField);
