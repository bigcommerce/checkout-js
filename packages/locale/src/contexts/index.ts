export { LocaleContext, LocaleContextType, useLocale } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
