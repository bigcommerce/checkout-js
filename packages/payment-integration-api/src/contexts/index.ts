export { CheckoutContext, CheckoutContextProps, useCheckout } from './checkout-context';
export {
    PaymentFormContextProps,
    PaymentFormContext,
    usePaymentFormContext,
} from './payment-form-context';
