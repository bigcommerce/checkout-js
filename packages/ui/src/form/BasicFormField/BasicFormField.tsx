import { Field, FieldConfig, FieldProps, getIn } from 'formik';
import { isDate, noop } from 'lodash';
import React, {
    Component,
    createElement,
    FunctionComponent,
    memo,
    useCallback,
    useMemo,
} from 'react';
import shallowEqual from 'shallowequal';

import { FormFieldContainer } from '../FormFieldContainer';

export interface BasicFormFieldProps extends FieldConfig {
    additionalClassName?: string;
    className?: string;
    testId?: string;
    onChange?(value: any): void;
}

type InnerFieldInputProps = FieldProps &
    Pick<FieldConfig, 'component' | 'render'> & {
        onChange?(value: string): void;
    };

type InnerFieldProps = Omit<BasicFormFieldProps, keyof FieldConfig> & InnerFieldInputProps;

class InnerFieldInput extends Component<InnerFieldInputProps> {
    componentDidUpdate({ field: prevField }: InnerFieldInputProps) {
        const {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            field: { value },
            onChange = noop,
        } = this.props;

        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const comparableValue = isDate(value) ? value.getTime() : value;
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const comparablePrevValue = isDate(prevField.value)
            ? prevField.value.getTime()
            : prevField.value;

        if (comparableValue !== comparablePrevValue) {
            onChange(value);
        }
    }

    render() {
        const { component = 'input', field, render } = this.props;

        if (render) {
            return render(this.props);
        }

        if (typeof component === 'string') {
            return createElement(component, field);
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-argument
        return createElement(component as any, this.props);
    }
}

const InnerField: FunctionComponent<InnerFieldProps> = memo(
    ({ additionalClassName, component, field, form, onChange, render, testId }) => {
        const input = useMemo(
            () => (
                <InnerFieldInput
                    component={component}
                    field={field}
                    form={form}
                    onChange={onChange}
                    render={render}
                />
            ),
            [field, form, onChange, component, render],
        );

        return (
            <FormFieldContainer
                additionalClassName={additionalClassName}
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                hasError={getIn(form.errors, field.name)}
                testId={testId}
            >
                {input}
            </FormFieldContainer>
        );
    },
    (
        { form: prevForm, field: prevField, ...prevProps },
        { form: nextForm, field: nextField, ...nextProps },
    ) =>
        shallowEqual(prevProps, nextProps) &&
        shallowEqual(prevForm, nextForm) &&
        shallowEqual(prevField, nextField),
);

const BasicFormField: FunctionComponent<BasicFormFieldProps> = ({
    additionalClassName,
    className,
    component,
    render,
    testId,
    onChange,
    ...rest
}) => {
    const renderInnerField = useCallback(
        (props: FieldProps) => (
            <InnerField
                {...props}
                additionalClassName={additionalClassName}
                className={className}
                component={component}
                onChange={onChange}
                render={render}
                testId={testId}
            />
        ),
        [additionalClassName, className, component, render, testId, onChange],
    );

    return <Field {...rest} render={renderInnerField} />;
};

export default memo(BasicFormField);
