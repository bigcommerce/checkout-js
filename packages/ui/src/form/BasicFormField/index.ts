export { default as BasicFormField } from './BasicFormField';
