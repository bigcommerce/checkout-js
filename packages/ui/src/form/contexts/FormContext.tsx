import { isFunction, noop } from 'lodash';
import React, { createContext, FunctionComponent, memo, ReactNode, useMemo, useState } from 'react';

export interface FormContextType {
    isSubmitted: boolean;
    setSubmitted(isSubmitted: boolean): void;
}

export const FormContext = createContext<FormContextType>({
    isSubmitted: false,
    setSubmitted: noop,
});

type FormContextFunction = (props: FormContextType) => ReactNode;

export interface FormProviderProps {
    initialIsSubmitted?: boolean;
    children: ReactNode | FormContextFunction;
}

export const FormProvider: FunctionComponent<FormProviderProps> = memo(
    ({ children, initialIsSubmitted = false }) => {
        const [isSubmitted, setSubmitted] = useState(initialIsSubmitted);
        const contextValue = useMemo(() => ({ isSubmitted, setSubmitted }), [isSubmitted]);

        return (
            <FormContext.Provider value={contextValue}>
                {isFunction(children) ? children({ isSubmitted, setSubmitted }) : children}
            </FormContext.Provider>
        );
    },
);

export default FormContext;
