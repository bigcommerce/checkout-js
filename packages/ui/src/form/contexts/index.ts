export { default as FormContext, FormContextType, FormProvider } from './FormContext';
