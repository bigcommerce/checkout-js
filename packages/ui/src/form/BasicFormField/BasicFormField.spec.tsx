import { mount, shallow } from 'enzyme';
import { Formik } from 'formik';
import { noop } from 'lodash';
import React from 'react';

import { FormContext } from '../contexts';

import BasicFormField from './BasicFormField';

describe('BasicFormField', () => {
    it('matches snapshot', () => {
        expect(shallow(<BasicFormField name="foobar" />)).toMatchSnapshot();
    });

    it('renders component with test ID', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: 'foobar' }}
                onSubmit={noop}
                render={() => <BasicFormField name="foobar" testId="test" />}
            />,
        );

        expect(component.find('.form-field').prop('data-test')).toBe('test');
    });

    it('changes appearance when there is error', async () => {
        const component = mount(
            <FormContext.Provider value={{ isSubmitted: true, setSubmitted: jest.fn() }}>
                <Formik
                    initialValues={{ foobar: '' }}
                    onSubmit={noop}
                    render={() => (
                        <BasicFormField
                            name="foobar"
                            render={({ field }) => <input {...field} type="text" />}
                            validate={() => 'Invalid'}
                        />
                    )}
                />
            </FormContext.Provider>,
        );

        component.find('input[name="foobar"]').simulate('change').simulate('blur');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        expect(component.find('.form-field').hasClass('form-field--error')).toBe(true);
    });

    it('renders input component by default', () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: '' }}
                onSubmit={noop}
                render={() => <BasicFormField name="foobar" />}
            />,
        );

        expect(component.exists('input')).toBe(true);
    });

    it('renders input component with date value', async () => {
        const component = mount(
            <Formik
                initialValues={{ foobar: '' }}
                onSubmit={noop}
                render={() => <BasicFormField name="foobar" />}
            />,
        );

        const date = new Date();

        component
            .find('input[name="foobar"]')
            .simulate('change', { target: { value: date, name: 'foobar' } })
            .simulate('blur');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        const input = component.find('input');

        expect(component.exists('input')).toBe(true);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        expect(input.get(0).props.value).toBe(date);
    });
});
