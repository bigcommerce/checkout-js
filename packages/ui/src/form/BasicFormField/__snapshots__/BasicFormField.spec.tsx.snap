// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicFormField matches snapshot 1`] = `
<FormikConnect(FieldInner)
  name="foobar"
  render={[Function]}
/>
`;
